### ISSUE TITLE
[ENH] avoid loading whole nifti-like image data when extracting time-series within a mask

### ISSUE THREAD

**Issue by MengxingLiu** [nilearn/nilearn#4811]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe your proposed enhancement in detail.

I am using nilearn to mask a large nifti using the following command:

```python
mask_nii = nib.load(mask)
self.mask_img = new_img_like(mask_nii, get_data(mask_nii)==ROI_id)
self.masker = NiftiMasker(mask_img=self.mask_img)
residual_img = nib.load(errts_file)
self.residual = self.masker.fit_transform(residual_img)
```
It looks like when masking nilearn loads the entire image into memory, which can cause memory issue when the residual image is big. I tracked the memory usage, it looks like it’s only at the step of

`self.residual = self.masker.fit_transform(residual_img) `
nilearn loads the residual image into memory.

Would it be possible to implement a feature that could control loading whole image or not when only needing to extract the time series within a small mask?

### Benefits to the change

It will decrease the memory usage and potentially allow more parallel jobs running at the same time

**Comment by bthirion** [comment]
I don't see how to do that with the masker we have.
If your image is 4D you can mask the 3D images sequentially to reduce memory load.

**Comment by man-shu** [comment]
Some more context:
We had this discussion with @MengxingLiu during the drop-in hour yesterday. We think it might be possible to use NifitLabelsMasker (where labels are basically several masks). However, while NifitLabelsMasker always aggregates values from voxels in a labeled region (via `strategy` parameter), @MengxingLiu would like the output to be available without any aggregation, meaning all voxel vaules should still be available after the transform.

One solution could be implementing a new strategy where no aggregation happens. NifitLabelsMasker uses a subset of the aggregation strategies from scipy ndimage: https://docs.scipy.org/doc/scipy/reference/ndimage.html. 

Possible solution:
[value_indices](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.value_indices.html#scipy.ndimage.value_indices) is one strategy I think could be used for this.

Possible caveat:
The output signal would be 3D instead of the usual 2D -- handling that might be an issue.

**Comment by MengxingLiu** [comment]
> I'm personally against having no aggregation in `NiftiLabelsMasker` ; this has [come up before](https://github.com/nilearn/nilearn/issues/2686) and been rejected (I think for good reasons). It also seems (to me) like it doesn't get to the core of the issue re : performance concerns.
> 
> I'd propose instead to subsample the mask (e.g., 4D to 3D as BT suggested). I'm also a little confused about this line:
> 
> ```
> self.mask_img = new_img_like(mask_nii, get_data(mask_nii)==ROI_id)
> ```
> 
> as it looks like this larger mask is already being subset by ROI, theoretically as `NiftiLabelsMasker` would do ? I do think we want to scale to larger data, as possible, so I'd be curious to know more about this exact use case !

Hi, thanks for checking this out.

NiftiLabelsMasker is averaging across voxels within the Mask. What I am trying to achieve here is to extract the time series within a mask without averaging, the shape of the matrix is expected to be number of scans by number of voxels.

**Comment by bthirion** [comment]
IIUC you can achieve that with the good old NiftiMasker.

**Comment by Remi-Gau** [comment]
- related 'old' issue: https://github.com/nilearn/nilearn/issues/719

**Comment by Remi-Gau** [comment]
- see also https://github.com/nilearn/nilearn/issues/1883

**Comment by Remi-Gau** [comment]
- and https://github.com/nilearn/nilearn/issues/1754

**Comment by emdupre** [comment]
Hi everyone,
Following up on a conversation at our core-dev meeting yesterday, I wanted to share some code from [`fmralign`](https://github.com/Parietal-INRIA/fmralign) that we use to access parcels in parallel. The important parts of the parcellation code are in [`ParcellationMasker`](https://github.com/Parietal-INRIA/fmralign/blob/main/fmralign/preprocessing.py#L20)

But I think we can do something even simpler in this usecase, by using [array proxies and memory-mapped access to the nii file](https://nipy.org/nibabel/images_and_memory.html) (since loading the functional file in memory multiple times seems to be the point-of-failure). If I understand the use case correctly, could you do something like this ?

```{python}
import numpy as np
import nibabel as nib

from nilearn import datasets, image

atlas_dataset = datasets.fetch_atlas_harvard_oxford('cort-maxprob-thr25-2mm')
atlas = atlas_dataset.maps
labels = atlas_dataset.labels

fmri_dataset = datasets.fetch_development_fmri(n_subjects=1, reduce_confounds=True)
fmri = nib.load(fmri_dataset.func[0])  # func still returns a list, even with one entry, so index
resample_atlas = image.resample_to_img(atlas, fmri, interpolation='nearest', copy_header=True, force_resample=True)

parcel_idx = np.unique(resample_atlas.dataobj)  # recheck labels since resampling
parcel_rois = [(resample_atlas.dataobj == idx) for idx in parcel_idx]

# note we don't need both of these outputs necessarily
# here we are generating a list of parcels with data corresponding to the labels array.
# this can be used directly (i.e., for signal processing), or, to get the data back into brain shape,
# we can use the empty array.
parcellated_img_data = np.empty(shape=fmri.dataobj.shape, dtype=np.float32)
parcel_data_list = []

for parcel in parcel_rois:  # this could also be done with joblib.Parallel for large N parcellations
    # avoid keeping in img array in internal cache
    # see : https://nipy.org/nibabel/images_and_memory.html#uncache-the-array
    parcel_data = np.asarray(fmri.dataobj)[parcel]
    
    # save out to both outputs, just for a complete example
    parcel_data_list.append(parcel_data)
    parcellated_img_data[parcel] = parcel_data

# convert everything back into one image
parcellated_fmri = fmri.__class__(parcellated_img_data, affine=fmri.affine, header=fmri.header, extra=fmri.extra)
```

Profiling with `tracemalloc`, this seems quite reasonable in terms of memory allocation :

```
Top 3 lines
#1: <ipython-input-87-0db149326663>:27: 193598.3 KiB
    parcel_data = np.asarray(fmri.dataobj)[parcel]
#2: nibabel/volumeutils.py:927: 193593.8 KiB
    arr = arr + inter1d
#3: <ipython-input-87-0db149326663>:21: 96797.0 KiB
    parcellated_img_data = np.empty(shape=fmri.dataobj.shape, dtype=np.float32)
388 other: 8928.2 KiB
Total allocated size: 492917.4 KiB
```

But if there are other painpoints of this use case I'm not aware of, please let me know !

**Comment by MengxingLiu** [comment]
Hi @emdupre 

Thank you very much for sharing the idea. I have tested it on my machine, it looks like if using joblib to parallize the jobs, the memory usage is higher than when it's single thread as for loop:

this is the code I tested on my machine. The first part is your code, the second part is using joblib to the same:

```
tracemalloc.start()
atlas_dataset = datasets.fetch_atlas_harvard_oxford('cort-maxprob-thr25-2mm')
atlas = atlas_dataset.maps
labels = atlas_dataset.labels

fmri_dataset = datasets.fetch_development_fmri(n_subjects=1, reduce_confounds=True)
fmri = nib.load(fmri_dataset.func[0])  # func still returns a list, even with one entry, so index
resample_atlas = image.resample_to_img(atlas, fmri, interpolation='nearest',force_resample=True)

parcel_idx = np.unique(resample_atlas.dataobj)  # recheck labels since resampling
parcel_rois = [(resample_atlas.dataobj == idx) for idx in parcel_idx]

# note we don't need both of these outputs necessarily
# here we are generating a list of parcels with data corresponding to the labels array.
# this can be used directly (i.e., for signal processing), or, to get the data back into brain shape,
# we can use the empty array.
parcellated_img_data = np.empty(shape=fmri.dataobj.shape, dtype=np.float32)
parcel_data_list = []

for parcel in parcel_rois:  # this could also be done with joblib.Parallel for large N parcellations
    # avoid keeping in img array in internal cache
    # see : https://nipy.org/nibabel/images_and_memory.html#uncache-the-array
    parcel_data = np.asarray(fmri.dataobj)[parcel]
    
    # save out to both outputs, just for a complete example
    parcel_data_list.append(parcel_data)
    parcellated_img_data[parcel] = parcel_data

# convert everything back into one image
parcellated_fmri = fmri.__class__(parcellated_img_data, affine=fmri.affine, header=fmri.header, extra=fmri.extra)
# confirm it matches the original data, so we were simply masking, as intended
np.allclose(fmri.dataobj, parcellated_fmri.dataobj)

snapshot = tracemalloc.take_snapshot()
display_top(snapshot)


from joblib import Parallel, delayed, dump, load

#tracemalloc.start()

# note we don't need both of these outputs necessarily
# here we are generating a list of parcels with data corresponding to the labels array.
# this can be used directly (i.e., for signal processing), or, to get the data back into brain shape,
# we can use the empty array.
parcellated_img_data = np.empty(shape=fmri.dataobj.shape, dtype=np.float32)

def extract_parcel_data(parcel):  # this could also be done with joblib.Parallel for large N parcellations
    # avoid keeping in img array in internal cache
    # see : https://nipy.org/nibabel/images_and_memory.html#uncache-the-array
    parcel_data = np.asarray(fmri.dataobj)[parcel]
    
    # save out to both outputs, just for a complete example
    #parcel_data_list.append(parcel_data)
    #parcellated_img_data[parcel] = parcel_data
    return parcel_data

n_jobs = len(parcel_rois)
tracemalloc.start()
parcel_data_list = Parallel(n_jobs=n_jobs)(delayed(extract_parcel_data)(parcel) for parcel in parcel_rois)

snapshot = tracemalloc.take_snapshot()
display_top(snapshot)

# convert everything back into one image
parcellated_fmri = fmri.__class__(parcellated_img_data, affine=fmri.affine, header=fmri.header, extra=fmri.extra)
# confirm it matches the original data, so we were simply masking, as intended
np.allclose(fmri.dataobj, parcellated_fmri.dataobj)


```

I allocated 30 cores and 8G memory to test the code, it is able to run through the for loop, but not the joblib parallel jobs. I did not find a very good way to monitor memory usage. tracemalloc was only able to track each single joblib job's usage so it's not hard to tell how much more memory is allocated when running multiple jobs simultaneously. 
```

(searchlight) [mliu23@login-prod-03 QUANTUM_4S]$ srun -p interactive,batch -c 30  --mem=8g --pty bash
srun: job 10430518 queued and waiting for resources
srun: job 10430518 has been allocated resources
[mliu23@p1cmp017 QUANTUM_4S]$ conda activate searchlight
(searchlight) [mliu23@p1cmp017 QUANTUM_4S]$ python3 mem_test.py
Top 3 lines
#1: QUANTUM_4S/mem_test.py:56: 193598.4 KiB
    parcel_data = np.asarray(fmri.dataobj)[parcel]
#2: nibabel/volumeutils.py:925: 193594.0 KiB
    arr = arr + inter1d
#3: QUANTUM_4S/mem_test.py:50: 96797.4 KiB
    parcellated_img_data = np.empty(shape=fmri.dataobj.shape, dtype=np.float32)
471 other: 8438.9 KiB
Total allocated size: 492428.6 KiB
Traceback (most recent call last):
  File "/cluster/home/mliu23/GIT/QUANTUM_4S/mem_test.py", line 95, in <module>
    parcel_data_list = Parallel(n_jobs=n_jobs)(delayed(extract_parcel_data)(parcel) for parcel in parcel_rois)
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 2007, in __call__
    return output if self.return_generator else list(output)
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 1650, in _get_outputs
    yield from self._retrieve()
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 1754, in _retrieve
    self._raise_error_fast()
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 1789, in _raise_error_fast
    error_job.get_result(self.timeout)
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 745, in get_result
    return self._return_or_raise()
  File "/cluster/tufts/halassalab/mliu23/condaenv/searchlight/lib/python3.9/site-packages/joblib/parallel.py", line 763, in _return_or_raise
    raise self._result
joblib.externals.loky.process_executor.TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

The exit codes of the workers are {SIGKILL(-9)}

```

My temporary solution for the need of accessing chunks of large image data and avoiding memory allocation is using shared memory between joblib jobs:
```

    # # Create shared memory
    #t.tic()
    # residual = residual_img.get_fdata()
    residual = np.load(f"{proc_dir}/errts.{sub_id}.npy")
    shm = shared_memory.SharedMemory(create=True, size=residual.nbytes)
    shared_residual = np.ndarray(residual.shape, dtype=residual.dtype, buffer=shm.buf)
    np.copyto(shared_residual, residual)

    print(f"running RDM {sub_id} session {ses_id}")

    results = Parallel(n_jobs=n_jobs)(delayed(run_jobs)(ROI_val,ROI_label, shared_residual) for ROI_val,ROI_label in zip(mask_csv["ROIid"],mask_csv["ROIname"]))

    # # Cleanup
    shm.close()
    shm.unlink()

```

**Comment by emdupre** [comment]
Hi @MengxingLiu , glad to hear that you've found something that works for your use case ! Just to clarify: 

> it looks like if using joblib to parallize the jobs, the memory usage is higher than when it's single thread as for loop:

This is expected, as you're now running multiple calls at the same time (in parallel). In your code, you're setting the `n_jobs` parameter to be equal to the number of parcels ; this will try to run all parcels at the same time as resources are available, and --- as you have 30 cores available --- is likely equal to setting `n_jobs=-1`. On an HPC (as it appears you are running this code), this is [generally not recommended](https://superuser.com/a/959103). I would select a much lower explicit value.

> tracemalloc was only able to track each single joblib job's usage so it's hard to tell how much more memory is allocated when running multiple jobs simultaneously.

Yes, `tracemalloc` is a Python library and is not intended for multithreaded processing. As you appear to be on an HPC, I would use the functionality available there ; e.g., `sacct` on a SLURM system. 

> My temporary solution for the need of accessing chunks of large image data and avoiding memory allocation is using shared memory between joblib jobs:

This works as well (!), though note that in this case you are loading the whole array in memory once and sharing it between processes, rather than avoiding loading and caching within each iteration, as I suggested in my snippet.

**Comment by emdupre** [comment]
I think this discussion has (already !) been very helpful re : expanding the documentation to cover use cases with high performance needs, would you agree @Remi-Gau @man-shu ?

**Comment by Remi-Gau** [comment]
Yes that cannot hurt.

But I think we need to start having some performance benchmark running regularly and some setup to help us do some profiling.

Am afraid that talking about or trying to optimize performance without measuring things may lead to focus on the wrong part of the code or may even makes things slower.
Also I'd like to have a way to flag a regression in performance the same way we have regression tests.

Some of those could also be part of our Doc.

See sklearn benchmark 

https://scikit-learn.org/scikit-learn-benchmarks/ 

See fmriprep

https://fmriprep.org/en/stable/benchmarks.html

https://fmriprep.org/en/stable/faq.html#how-much-cpu-time-and-ram-should-i-allocate-for-a-typical-fmriprep-run

**Comment by emdupre** [comment]
> But I think we need to start having some performance benchmark running regularly and some setup to help us do some profiling.

> Am afraid that talking about or trying to optimize performance without measuring things may lead to focus on the wrong part of the code or may even makes things slower.
Also I'd like to have a way to flag a regression in performance the same way we have regression tests.

Hm. I don't know that I see how necessary this really is for the need we had discussed in the last core dev call.

Maybe worth opening a new issue to discuss that specifically ? It sounds like the original usage issue that this was intended to address is resolved, and we've cycled through a few topics here already that make it hard to linearly follow.

**Comment by Remi-Gau** [comment]
yes sorry you are correct this is pretty much 99% unrelated: just saw the word performance and my brain went on a side track
